alacritty/src/display/hint.rs

```diff
@@ -258,30 +381,65 @@ pub fn highlighted_at<T>(
         }
 
-        hint.regex.with_compiled(|regex| {
-            // Setup search boundaries.
-            let mut start = term.line_search_left(point);
-            start.line = max(start.line, point.line - MAX_SEARCH_LINES);
-            let mut end = term.line_search_right(point);
-            end.line = min(end.line, point.line + MAX_SEARCH_LINES);
+        if let Some((hyperlink, bounds)) =
+            hint.content.hyperlinks.then(|| hyperlink_at(term, point)).flatten()
+        {
+            return Some(HintMatch::Hyperlink { bounds, action: hint.action.clone(), hyperlink });
+        }
 
-            // Function to verify that the specified point is inside the match.
-            let at_point = |rm: &Match| *rm.end() >= point && *rm.start() <= point;
+        if let Some(bounds) = hint.content.regex.as_ref().and_then(|regex| {
+            regex.with_compiled(|regex| regex_match_at(term, point, regex, hint.post_processing))
+        }) {
+            return Some(HintMatch::Regex { bounds, action: hint.action.clone() });
+        }
 
-            // Check if there's any match at the specified point.
-            let mut iter = RegexIter::new(start, end, Direction::Right, term, regex);
-            let regex_match = iter.find(at_point)?;
+        None
+    })
+}
 
-            // Apply post-processing and search for sub-matches if necessary.
-            let regex_match = if hint.post_processing {
-                HintPostProcessor::new(term, regex, regex_match).find(at_point)
-            } else {
-                Some(regex_match)
-            };
+/// Retrive the hyperlink with its range, if there is one at the specified point.
+fn hyperlink_at<T>(term: &Term<T>, point: Point) -> Option<(Hyperlink, Match)> {
+    let hyperlink = term.grid()[point].hyperlink()?;
+
+    let viewport_start = Line(-(term.grid().display_offset() as i32));
+    let viewport_end = viewport_start + term.bottommost_line();
+
+    let mut match_start = Point::new(point.line, Column(0));
+    let mut match_end = Point::new(point.line, Column(term.columns() - 1));
+    let grid = term.grid();
+
+    // Find start and end line containing the same hyperlink inside viewport that is adjustsent to
+    // the requested point's line.
+
+    // Find start.
+    loop {
+        let next_line = match_start.line - 1i32;
+        if next_line < viewport_start
+            || !grid[next_line]
+                .into_iter()
+                .any(|cell| cell.hyperlink().map(|h| h == hyperlink).unwrap_or(false))
+        {
+            break;
+        }
 
-            regex_match.map(|bounds| HintMatch { action: hint.action.clone(), bounds })
-        })
-    })
+        match_start.line = next_line;
+    }
+
+    // Find end.
+    loop {
+        let next_line = match_end.line + 1i32;
+        if next_line >= viewport_end
+            || !grid[next_line]
+                .into_iter()
+                .any(|cell| cell.hyperlink().map(|h| h == hyperlink).unwrap_or(false))
```

Comment:
It's not immediately clear what this does, should be documented, probably extracted in a separate variable with a nice name.